### ISSUE TITLE
[Table] Support drag and drop of rows within the table

### ISSUE THREAD

**Issue by gillibrand** [carbon-design-system/carbon-addons-iot-react#3751]
### What package is this for?

React

### Summary

Add drag and drop support to the table component. Specially, it should be possible to drag rows of a table and drop them on other rows of the same table, i.e., moving "file" rows into "folder" rows. 

The enhancement is needed by the ELM team to move an existing UI with this feature over to the PAL table. Minimally this should:

1. Support showing a drag handle on a row, ideally in the "first" column there checkmarks are shown.
2. Allow a a row to be dragged from its handle and move freely around. 
3. Show a caller provided drag image during the drag to represent the data on the row (showing the full row is normally too big).
4. Support dragging any number of selected rows if there is a selection.
5. While dragging, highlight droppable rows based on the [Carbon DnD guidance](https://pages.github.ibm.com/cdai-design/pal/patterns/drag-and-drop/usage/).
6. Support dropping on any row in the same table. The caller must provide a list of row IDs that can accept the droppable.
7. Dragging between tables and windows is NOT required at this time.
8. On drop, fire a callback describing the action so the caller can respond to it.

This _could_ be used for reordering (assuming the table is unsorted) as described here, but the UI would make it appear that you are dropping row ON to other rows, not between them. That's both misleading and means it may not be possible to reorder something to the top the table (would dropping on the first row add the dropped row before or after?). That's not the use case this is addressing at this time.

#### Example 

Here's a shot of what the table will look like when dragging over an acceptable row.

![image](https://user-images.githubusercontent.com/805573/225446380-39621d08-fdbe-4e7f-a07e-3cbd907dd786.png)

### Specific timeline issues / requests

ASAP or course. Within a month is ideal.

### Available extra resources

I, Jay, am available to work on this. Never having contributed to PAL before, I'm sure I'll need some coaching, but the main code changes I should be able to handle.

**Comment by gillibrand** [comment]
@davidicus I started looking into this. Note that the requirements are very specific to the ELM use case. I saw there was a roadmap item for "DnD row reordering", but this is not that.

PAL uses the "React DnD" 3rd party library already in at least one place. I experimented with that, but ran into some limitations:

1. Would require Carbon changes to expose refs on some table components.
2. The normal way to show a drag image/preview requires generating them before the drag starts. The PAL reorderable List component does this and hides them behind the list itself until needed, but I worry about performance of that for large tables.
3. Use html5 native DnD events. Those give almost no control over the cursor, so it won't match the grab/grabbing cursor 	Carbon recommendation. I also personally hate ghostbusters "no" cursor it constantly flashes during a drag. (There is an npm package that supplies alternative engine as a possible work around).

Because of those I'm inclined to just do this with custom mousemove events. I prototyped that and it's really not too much code and gives a very tailored solution. Of course, native drag events have the huge advantage that they can cross between windows. It would be very cool to drag a table from from one window to another. That's more complicated though, and not a requirement currently. I've asked the product manager for clarification if that's a hidden expectation or on the roadmap though.

For changes in the PAL Table, I think this would work similar to the `hasRowSelection` prop, which add extra `<th>` into the table head (for this it would be empty, but just take up space), and a leading `<td>` in the rows with a drag handle (6 dots) icon. That would show before the selection checkbox/radio if both are enabled. I haven't looked into expandable rows yet.

The table would likely also need new props like:

1. `onDragStarted` callback to pass the caller the dragged row and have it set the next prop...
2. `droppableRowIds` to know which rows to highlight when dragging over them.
3. `onDrop` callback that passed the row ID and/or the rows values.
